README.md
# pico8-custom-editor README

This is the README for your extension "pico8-custom-editor". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "pico8-custom-editor" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

media/base.js
import {html, render} from 'https://unpkg.com/lit-html?module';

(function() {

    let vscode;
    try{
        vscode = acquireVsCodeApi();
    } catch {}


const testData = [
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '111111112222222233333333444444445555555566666666777777778888888899999999aaaaaaaabbbbbbbbccccccccddddddddeeeeeeeeffffffff00000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000004444000044440000444400000000000000000000444400000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000443f00004444000043f30000000000000000000043f300000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000ffff0000ffff0000ffff00000000000000000000ffff00000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000001111000011110000111100000000000000000000111100000000000000000000000000000000000000000000000000000000000000000000000000',
    '000000000f5555f00f5555f00f5555f000000000000000000f5555f0000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000d00d0000d00d0000d00d00000000000000000000d00d00000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000004000044000000004400000000000000000000000040000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000004444000044440000444400000000000000000000444400000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000443f00004444000043f30000000000000000000043f300000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000ffff0000ffff0000ffff00000000000000000000ffff00000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000001111000f1111000f111100000000000000000000111100000000000000000000000000000000000000000000000000000000000000000000000000',
    '0000000000f55500005555000055550000000000000000000f5555f0000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000dd000005000f0005000f0000000000000000000d00d00000000000000000000000000000000000000000000000000000000000000000000000000',
    '000000000000000000d0000000d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '22222222222222222222222220002021120200020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '22000000000000000000002202000202202000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20200002000000002000020200200021120002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020200002002000020200200020202202020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20002002200000022002000220002202202200020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020220020220200220200202022220122220200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000222202002022220000220200222222002020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20202022121221212202020212122121120221210000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20002021111111111202000212122021121221210000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000202111111112020000220200222222002020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020021111111111200200202022221022220200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000202111111112020000220002202202200020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000202111111112020000200020202202020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020021111111111200200200200021120002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000202111111112020000202000202202000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20002021111111111202000220002021120200020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20202022121221212202020200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20000222202002022220000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020220020220200220200200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20002002200000022002000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20020200002002000020200200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '20200002000000002000020200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '22000000000000000000002200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    '22222222222222222222222200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000']

    const state = {
        activeColor: 'black',
        activeSprite: 0,
        spriteCtx: undefined,
        previewCtx: undefined,
        pixelData: testData,
        page: 0,
        oldMessage: undefined
    }

    const colorHexs = [
        "#000000",
        "#1D2B53",
        "#7E2553",
        "#008751",
        "#AB5236",
        "#5F574F",
        "#C2C3C7",
        "#FFF1E8",
        "#FF004D",
        "#FFA300",
        "#FFEC27",
        "#00E436",
        "#29ADFF",
        "#83769C",
        "#FF77A8",
        "#FFCCAA",
    ]

    window.addEventListener('message', event => {
        const documentText = event.data.text;
        const lines = documentText.split('\n');
        let found = false;
        let gfxLines = [];
        lines.forEach(line => {
            if(found){
                if(line.length < 128) {
                    found = false;
                } else {
                    gfxLines.push(line);
                }
            } 
            if(line.includes('__gfx__')) {
                found = true
            }

        });
        if(gfxLines.length === 0){
            state.pixelData = new Array(64).fill(new Array(128).fill(0).join(''))
            event.data.text += '__gfx__\n'
            state.pixelData.forEach( row => event.data.text += row + '\n');
            // fill in the gfx section
        } else {
            state.pixelData = gfxLines;

        }
        updatePreview();
        updateCanvas();
    });

    const colors = ['black', 'darkblue', 'purple', 'darkgreen', 'brown',
                    'darkgrey', 'grey', 'white', 'red', 'orange', 'yellow',
                    'green', 'blue', 'indigo', 'pink', 'peach'];

    

    const isSelected = (a, b) => a === b ? 'selected' : '';

    const update = () => html`
        <div class="center-full">
            <div class="topbar"></div>
            <div class="viewContainer" @mousemove=${handleDraw} @click=${handleDraw}>
                <canvas class="spriteView"></canvas>
                <div class="colorView" @mousemove=${handleColor} @click=${handleColor}>
                    ${ colors.map(color => html`<div class="color ${color} ${isSelected(color, state.activeColor)}" color="${color}"></div>`) }
                </div>
            </div>
            <div class="sprites">
                <canvas class="spriteContainer"></canvas>
                <div class="spriteContainer highlighter" @mousemove=${handleSprite} @click=${handleSprite}>
                    ${ (new Array(64).fill(0).map( (_,i) => html`<div sprite="${i}" class="sprite ${isSelected(state.activeSprite, i)}"></div>`) )}
                </div>
            </div>
            <div class="topbar"></div>
        </div>
    `

    // CANVAS

    const updateCanvas = () => {
    
        // draw the sprite from the data
        let x = (state.activeSprite % 16) * 8;
        let y = Math.floor((state.activeSprite / 16)) * 8;
        let pixels = [];
        for(let i = 0; i < 8; i++){
            pixels.push(state.pixelData[y + i].substring(x, x + 8));
        }

        let ctx = state.spriteCtx;
        ctx.canvas.width = ctx.canvas.scrollWidth;
        ctx.canvas.height = ctx.canvas.scrollHeight;

        ctx.fillStyle = colorHexs[0];
        ctx.fillRect(0,0, ctx.canvas.scrollWidth, ctx.canvas.scrollHeight);

        let width = Math.round(ctx.canvas.scrollWidth / 8);
        let height = Math.round(ctx.canvas.scrollHeight / 8);
        pixels.forEach((row, y) => row.split('').forEach( (cell, x) => {
            let color = parseInt(cell, 16);
            ctx.fillStyle = colorHexs[color];
            ctx.strokeStyle = colorHexs[color];
            ctx.fillRect(x * width, y * height, width, height);
        }))
    }

    const updatePreview = () => {
        const ctx = state.previewCtx;
        //ctx.fillStyle = colorHexs[0];
        ctx.canvas.width = ctx.canvas.scrollWidth;
        ctx.canvas.height = ctx.canvas.scrollHeight;

        let example = document.querySelector('.sprite');
        const stringSize = getComputedStyle(example)['width']
        const width = parseFloat(stringSize.substring(0, stringSize.length - 2)) / 8;
        const height = parseFloat(stringSize.substring(0, stringSize.length - 2)) / 8;

        // for (let y = 0; y < 4; y++){
        //     for (let x = 0; x < 16; x++){
        //         ctx.beginPath();
        //         ctx.strokeStyle = '#ff0000';
        //         ctx.lineWidth = 3;
        //         ctx.rect(x * width, y * height, width, height)
        //         ctx.stroke()
        //     }
        // }

        const currentPage = (state.page * 4)
        for(let y = currentPage; y < (currentPage + (8*4)); y++){
            let row = state.pixelData[y]
            row.split('').forEach((cell,x) => {
                let color = parseInt(cell, 16);
                ctx.fillStyle = colorHexs[color];
                ctx.strokeStyle = colorHexs[color];
                ctx.fillRect(Math.floor(x * width), Math.floor(y * height), Math.ceil(width), Math.ceil(height));
            })
        }
    }

    // EVENTS

    const handleColor = (e) => {
        if( e.buttons == 1 || e.type === 'click' ){
            const color = e.target.getAttribute('color');
            if(color !== null) {
                state.activeColor = color;
                render(update(), document.body);
            }
        }   
    }

    const handleSprite = (e) => {
        if( e.buttons == 1 || e.type === 'click'){
            const sprite = e.target.getAttribute('sprite');
            if(sprite !== null){
                state.activeSprite = Number(sprite);
                updateCanvas();
                render(update(), document.body);
            }
        }
    }

    const isSameMessage = (a, b) => 
            a !== undefined &&
            b !== undefined &&
            a.type === b.type && 
            a.pos.x === b.pos.x &&
            a.pos.y === b.pos.y &&
            a.sprite === b.sprite &&
            a.color === b.color;

    const handleDraw = (e) => {
        if( e.buttons == 1 || e.type === 'click'){
            const stringSize = getComputedStyle(e.target)['width']
            let size = parseFloat(stringSize.substring(0,stringSize.length - 2)) / 8;
            let x = Math.floor((e.clientX - state.spriteCtx.canvas.offsetLeft) / size);
            let y = Math.floor((e.clientY - state.spriteCtx.canvas.offsetTop) / size);
            const newMessage = { 
                type: 'draw', 
                pos: {x, y},
                sprite: state.activeSprite,
                color: colors.indexOf(state.activeColor).toString(16)
            };
            if(x >= 0 && x < 16 && y >= 0 && y < 16 && !isSameMessage(state.oldMessage, newMessage)) {
                try{
                    console.log('post message')
                    vscode.postMessage(newMessage);
                    state.oldMessage = newMessage;
                } catch {}
            }
        }
    }

    render(update(), document.body);

    state.spriteCtx = document.querySelector('.spriteView').getContext("2d");

    state.previewCtx = document.querySelector('.spriteContainer').getContext("2d");

    updateCanvas();
    updatePreview();

})();

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import * as path from 'path';

export class Pico8SpriteEditor implements vscode.CustomTextEditorProvider {

	public static register(context: vscode.ExtensionContext): vscode.Disposable {
		const provider = new Pico8SpriteEditor(context);
		const providerRegistration = vscode.window.registerCustomEditorProvider(Pico8SpriteEditor.viewType, provider);
		return providerRegistration;
	}

	private static readonly viewType = 'advanderar.customeditor8';

	constructor(
		private readonly context: vscode.ExtensionContext
	) { 
		console.log('Constructor!');
	}

	resolveCustomTextEditor(document: vscode.TextDocument, webviewPanel: vscode.WebviewPanel, token: vscode.CancellationToken): void | Thenable<void> {
		
		webviewPanel.webview.options = {
			enableScripts: true,
		};
		webviewPanel.webview.html = this.getHtmlForWebview(webviewPanel.webview);

		function updateWebview() {
			webviewPanel.webview.postMessage({
				type: 'update',
				text: document.getText(),
			});
		}

		const changeDocumentSubscription = vscode.workspace.onDidChangeTextDocument(e => {
			if (e.document.uri.toString() === document.uri.toString()) {
				updateWebview();
			}
		});

		// Make sure we get rid of the listener when our editor is closed.
		webviewPanel.onDidDispose(() => {
			changeDocumentSubscription.dispose();
		});

		// Receive message from the webview.
		webviewPanel.webview.onDidReceiveMessage(e => {
			switch(e.type){
				case 'draw':
					this.updateDocument(document, e);
					break;
				default:
					break;
			}
		});

		updateWebview();
	}

	getHtmlForWebview(webview: vscode.Webview) : string {

		const scriptUri = webview.asWebviewUri(vscode.Uri.file(
			path.join(this.context.extensionPath, 'media', 'base.js')
		));

		const styleUri = webview.asWebviewUri(vscode.Uri.file(
			path.join(this.context.extensionPath, 'media', 'style.css')
		));

		// console.log(scriptUri);

		return `
			<!DOCTYPE html>
			<html lang="en" style="padding: 0; margin: 0; background-color: var(--dark-grey);">
				<head>
					<link href="${styleUri}" rel="stylesheet" />
				</head>
				<body style="margin: 0; padding: 0;"></body>
				<script type="module" src="${scriptUri}"></script>
			</html>
		`;
	}

	updateDocument(document: vscode.TextDocument, change: any) {
		const headerLength = 8;
		let text = document.getText();
		if(text.indexOf('__gfx__') === -1) {
			text += '__gfx__\n';
			text += (new Array(64).fill(new Array(128).fill(0).join(''))).join('\n')
		}
		let graphicsAndRest = text.substr(text.indexOf('__gfx__') + headerLength);

		const row = ((change.sprite % 16) * 8) + change.pos.x;
		const column = (Math.floor((change.sprite / 16)) * 8) + change.pos.y;
		if( row < 0 || column < 0) return;
		console.log(row, column);
		const arrayGraphics = graphicsAndRest.split('\n').map(row => row.split(''));
		arrayGraphics[column][row] = change.color;

		graphicsAndRest = arrayGraphics.map( row => row.join('')).join('\n');
		
		const replacement = text.substr(0, text.indexOf('__gfx__') + 8) + graphicsAndRest;

		const edit = new vscode.WorkspaceEdit();
		edit.replace(document.uri, new vscode.Range(0, 0, document.lineCount, 0), replacement)

		return vscode.workspace.applyEdit(edit);
	}
}

export function activate(context: vscode.ExtensionContext) {
	// Register our custom editor providers
	context.subscriptions.push(Pico8SpriteEditor.register(context));
}

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = exports.Pico8SpriteEditor = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const path = require("path");
class Pico8SpriteEditor {
    constructor(context) {
        this.context = context;
        console.log('Constructor!');
    }
    static register(context) {
        const provider = new Pico8SpriteEditor(context);
        const providerRegistration = vscode.window.registerCustomEditorProvider(Pico8SpriteEditor.viewType, provider);
        return providerRegistration;
    }
    resolveCustomTextEditor(document, webviewPanel, token) {
        webviewPanel.webview.options = {
            enableScripts: true,
        };
        webviewPanel.webview.html = this.getHtmlForWebview(webviewPanel.webview);
        function updateWebview() {
            webviewPanel.webview.postMessage({
                type: 'update',
                text: document.getText(),
            });
        }
        const changeDocumentSubscription = vscode.workspace.onDidChangeTextDocument(e => {
            if (e.document.uri.toString() === document.uri.toString()) {
                updateWebview();
            }
        });
        // Make sure we get rid of the listener when our editor is closed.
        webviewPanel.onDidDispose(() => {
            changeDocumentSubscription.dispose();
        });
        // Receive message from the webview.
        webviewPanel.webview.onDidReceiveMessage(e => {
            switch (e.type) {
                case 'draw':
                    this.updateDocument(document, e);
                    break;
                default:
                    break;
            }
        });
        updateWebview();
    }
    getHtmlForWebview(webview) {
        const scriptUri = webview.asWebviewUri(vscode.Uri.file(path.join(this.context.extensionPath, 'media', 'base.js')));
        const styleUri = webview.asWebviewUri(vscode.Uri.file(path.join(this.context.extensionPath, 'media', 'style.css')));
        // console.log(scriptUri);
        return `
			<!DOCTYPE html>
			<html lang="en" style="padding: 0; margin: 0; background-color: var(--dark-grey);">
				<head>
					<link href="${styleUri}" rel="stylesheet" />
				</head>
				<body style="margin: 0; padding: 0;"></body>
				<script type="module" src="${scriptUri}"></script>
			</html>
		`;
    }
    updateDocument(document, change) {
        const headerLength = 8;
        let text = document.getText();
        if (text.indexOf('__gfx__') === -1) {
            text += '__gfx__\n';
            text += (new Array(64).fill(new Array(128).fill(0).join(''))).join('\n');
        }
        let graphicsAndRest = text.substr(text.indexOf('__gfx__') + headerLength);
        const row = ((change.sprite % 16) * 8) + change.pos.x;
        const column = (Math.floor((change.sprite / 16)) * 8) + change.pos.y;
        if (row < 0 || column < 0)
            return;
        console.log(row, column);
        const arrayGraphics = graphicsAndRest.split('\n').map(row => row.split(''));
        arrayGraphics[column][row] = change.color;
        graphicsAndRest = arrayGraphics.map(row => row.join('')).join('\n');
        const replacement = text.substr(0, text.indexOf('__gfx__') + 8) + graphicsAndRest;
        const edit = new vscode.WorkspaceEdit();
        edit.replace(document.uri, new vscode.Range(0, 0, document.lineCount, 0), replacement);
        return vscode.workspace.applyEdit(edit);
    }
}
exports.Pico8SpriteEditor = Pico8SpriteEditor;
Pico8SpriteEditor.viewType = 'advanderar.customeditor8';
function activate(context) {
    // Register our custom editor providers
    context.subscriptions.push(Pico8SpriteEditor.register(context));
}
exports.activate = activate;
//# sourceMappingURL=extension.js.map